src/main.c

```diff
@@ -124,44 +128,40 @@ static void configure_goal(void)
 		speaker_play_competition();
 		break;
 	}
-}
-
-/**
- * @brief Execute the exploration phase.
- */
-static void exploration_phase(void)
-{
-	set_search_initial_direction(NORTH);
-	configure_goal();
 	set_target_goal();
-
-	movement_phase(false);
-
-	set_run_sequence();
-	save_maze();
+	configure_speed(false);
 }
 
 /**
- * @brief Configure the robot and execute the search/run.
+ * @brief Let the user reuse the saved maze for a run or discard it.
  */
-static void configure_and_move(void)
+static void configure_reuse(void)
 {
-	if (!maze_is_saved())
-		exploration_phase();
-	led_bluepill_on();
 	switch (button_user_wait_action()) {
 	case BUTTON_SHORT:
 		load_maze();
+		configure_speed(true);
 		break;
 	case BUTTON_LONG:
 		reset_maze();
 		led_bluepill_off();
-		exploration_phase();
-		led_bluepill_on();
+		configure_goal();
 		break;
 	}
-	while (1)
-		movement_phase(true);
+}
+
+/**
+ * @brief Start the robot configuration process for an exploration/run.
+ */
+static void configure_start(void)
+{
+	if (maze_is_saved()) {
+		led_bluepill_on();
+		configure_reuse();
```

Comment:
The configure_reuse function includes configure_goal, that includes configure_speed, it is hard to understand the paths. Could you group on a different manner?

Reply:
Replying to this and the previous comments:

I disagree. I think this implementation is easier to follow (although different, and that may be the cause why it feels harder to follow at first, since you are probably used to the previous code).

There is a huge improvement with this PR: this setup/configuration code is documented with a graph. If you have not tried to generate the documentation locally, here is the result:

![graphviz-927c595d49bbfda9cd978ebf7b9de8e960ed5ff6](https://user-images.githubusercontent.com/526577/70864606-f1efe980-1f53-11ea-9629-0c36befe827b.png)

This is objectively simpler than the previous implementation.

If you can suggest a simpler configuration procedure, I will change it of course. But cannot think of a better way right now. :innocent:

Reply:
I like a lot the configuration and the user state machine :), the comments were related to the code, or how that is implemented (functions names and dependencies)